src/pubsub/src/publisher/publisher.rs

```diff
@@ -216,50 +275,74 @@ impl Worker {
+    ///    send tasks that have completed, preventing the set from growing indefinitely.
+    ///
+    /// The loop terminates when the `rx` channel is closed, which happens when all
+    /// `Publisher` clones have been dropped.
     async fn run(mut self) {
         let mut batch = Batch::new();
         let delay = self.batching_options.delay_threshold;
         let message_limit = self.batching_options.message_count_threshold;
-
-        let flush = |b: Batch| {
-            tokio::spawn(b.send(self.client.clone(), self.topic_name.clone()));
-        };
+        let mut inflight = FuturesUnordered::new();
 
         let timer = tokio::time::sleep(delay);
         // Pin the timer to the stack.
         tokio::pin!(timer);
         loop {
             tokio::select! {
+                // Remove finished futures from the inflight messages.
+                _ = inflight.next(), if !inflight.is_empty() => {},
                 // Handle timer events.
                 // This branch will only be checked when there is a non-empty batch,
                 // so this will not fire continuously.
                 _ = &mut timer, if !batch.is_empty() => {
-                    flush(std::mem::take(&mut batch));
+                    batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
                 }
                 // Handle receiving a message from the channel.
                 msg = self.rx.recv() => {
                     match msg {
-                        Some(msg) => {
+                        Some(ToWorker::Publish(msg)) => {
                             // Reset the timer if this is the first message to be added to the batch.
                             if batch.is_empty() {
                                 timer.as_mut().reset(tokio::time::Instant::now() + delay);
                             }
                             batch.push(msg);
                             if batch.len() as u32 >= message_limit {
-                                flush(std::mem::take(&mut batch));
+                                batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
                             }
                         },
+                        Some(ToWorker::Flush(tx)) => {
+                            batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
+                            // Wait on all the tasks that exist right now.
+                            // We could instead tokio::spawn this as well so the publisher
+                            // can keep working on additional messages. The worker would
+                            // also need to keep track of any pending flushes, and make sure
+                            // all of those resolve as well.
+                            let mut flushing = std::mem::take(&mut inflight);
+                            while flushing.next().await.is_some() {}
```

Comment:
I spent too much time thinking about the tradeoffs between this vs. `futures::future::join_all()`.  Instead of wasting more of your time: please consider that as an alternative. If you did already, let me know why this is better?